From: Chad Monroe <chad@monroe.io>
Date: Tue, 28 Oct 2025 16:14:18 -0700
Subject: [PATCH] wifi: mt76: fix multi-radio on-channel scanning

avoid unnecessary channel switch when performing an on-channel scan
using a multi-radio device.

Fixes: c56d6edebc1f ("wifi: mt76: mt7996: use emulated hardware scan support")
Signed-off-by: Chad Monroe <chad@monroe.io>
---
 mt76.h |    1 +
 scan.c |   23 +++++++++++++++++------
 2 files changed, 18 insertions(+), 6 deletions(-)

--- a/mt76.h
+++ b/mt76.h
@@ -1021,6 +1021,7 @@ struct mt76_dev {
 		struct mt76_vif_link *mlink;
 		struct mt76_phy *phy;
 		int chan_idx;
+		bool offchannel;
 	} scan;
 
 #ifdef CONFIG_NL80211_TESTMODE
--- a/scan.c
+++ b/scan.c
@@ -16,7 +16,7 @@ static void mt76_scan_complete(struct mt
 
 	clear_bit(MT76_SCANNING, &phy->state);
 
-	if (dev->scan.chan && phy->main_chandef.chan &&
+	if (dev->scan.offchannel && dev->scan.chan && phy->main_chandef.chan &&
 	    !test_bit(MT76_MCU_RESET, &dev->phy.state))
 		mt76_set_channel(phy, &phy->main_chandef, false);
 	mt76_put_vif_phy_link(phy, dev->scan.vif, dev->scan.mlink);
@@ -87,6 +87,7 @@ void mt76_scan_work(struct work_struct *
 	struct cfg80211_chan_def chandef = {};
 	struct mt76_phy *phy = dev->scan.phy;
 	int duration = HZ / 9; /* ~110 ms */
+	bool onchannel = false;
 	int i;
 
 	if (dev->scan.chan_idx >= req->n_channels) {
@@ -94,28 +95,38 @@ void mt76_scan_work(struct work_struct *
 		return;
 	}
 
-	if (dev->scan.chan && phy->num_sta) {
+	if (dev->scan.chan && dev->scan.offchannel && phy->num_sta) {
 		dev->scan.chan = NULL;
 		mt76_set_channel(phy, &phy->main_chandef, false);
 		goto out;
 	}
 
 	dev->scan.chan = req->channels[dev->scan.chan_idx++];
-	cfg80211_chandef_create(&chandef, dev->scan.chan, NL80211_CHAN_HT20);
-	mt76_set_channel(phy, &chandef, true);
+	if (phy->main_chandef.chan &&
+	    phy->main_chandef.chan->center_freq == dev->scan.chan->center_freq)
+		onchannel = true;
+
+	if (onchannel)
+		chandef = phy->main_chandef;
+	else
+		cfg80211_chandef_create(&chandef, dev->scan.chan, NL80211_CHAN_HT20);
+
+	dev->scan.offchannel = !onchannel;
+	mt76_set_channel(phy, &chandef, !onchannel);
 
 	if (!req->n_ssids ||
 	    chandef.chan->flags & (IEEE80211_CHAN_NO_IR | IEEE80211_CHAN_RADAR))
 		goto out;
 
-	duration = HZ / 16; /* ~60 ms */
+	if (dev->scan.offchannel)
+		duration = HZ / 16; /* ~60 ms */
 	local_bh_disable();
 	for (i = 0; i < req->n_ssids; i++)
 		mt76_scan_send_probe(dev, &req->ssids[i]);
 	local_bh_enable();
 
 out:
-	if (dev->scan.chan)
+	if (dev->scan.chan && dev->scan.offchannel)
 		duration = max_t(int, duration,
 			         msecs_to_jiffies(req->duration +
 						  (req->duration >> 5)));
